lib/msf/core/payload/adapter/fetch.rb

```diff
@@ -0,0 +1,316 @@
+  # in Framework, and if the underlying payload type/host/port are the same, the URI
+  # will be, too.
+  #
+  def default_srvuri
+    # If we're in framework, payload is in datastore; msfvenom has it in refname
+    payload_name = datastore['payload'] ||= refname
+    decoded_uri = payload_name.dup
+    # there may be no transport, so leave the connection string off if that's the case
+    if module_info['ConnectionType'].upcase == 'REVERSE'
+      decoded_uri << "[#{datastore['LHOST']}]" unless datastore['LHOST'].nil?
+      decoded_uri << datastore['LPORT'].to_s unless datastore['LPORT'].nil?
+    end
+    if module_info['ConnectionType'].upcase == 'BIND'
+      decoded_uri << "[#{datastore['LHOST']}]" unless datastore['LHOST'].nil?
+      decoded_uri << datastore['RPORT'].to_s unless datastore['RPORT'].nil?
+    end
+    Base64.urlsafe_encode64(OpenSSL::Digest::MD5.new(decoded_uri).digest, padding: false)
+  end
+
+  def download_uri
+    "#{srvhost}:#{srvport}/#{srvuri}"
+  end
+
+  def fetch_bindhost
+    datastore['FetchListenerBindAddress'].blank? ? srvhost : datastore['FetchListenerBindAddress']
+  end
+
+  def fetch_bindport
+    datastore['FetchListenerBindPort'].blank? ? srvport : datastore['FetchListenerBindPort']
+  end
+
+  def generate(opts = {})
+    vprint_status('Generate')
+    opts[:arch] ||= module_info['AdaptedArch']
+    opts[:code] = super
+    @srvexe = generate_payload_exe(opts)
+    cmd = generate_fetch_commands
+    vprint_status("#{cmd}")
+    cmd
+  end
+
+  def generate_fetch_commands
+    vprint_status('generate_fetch_commands')
+    # TODO: Make a check method that determines if we support a platform/server/command combination
+    #
+    case datastore['FETCH_COMMAND'].upcase
+    when 'FTP'
+      vprint_status("Dispatching to FTP")
+      return _generate_ftp_command
+    when 'TNFTP'
+      vprint_status("Dispatching to TNFTP")
+      return _generate_tnftp_command
+    when 'WGET'
+      vprint_status("Dispatching to WGET")
+      return _generate_wget_command
+    when 'CURL'
+      vprint_status("Dispatching to CURL")
+      return _generate_curl_command
+    when 'TFTP'
+      vprint_status("Dispatching to TFTP")
+      return _generate_tftp_command
+    when 'CERTUTIL'
+      vprint_status("Dispatching to CERTUTIL")
+      return _generate_certutil_command
+    else
+      fail_with(Msf::Module::Failure::BadConfig, 'Unsupported Binary Selected')
+    end
+  end
+
+  def generate_stage(opts = {})
+    opts[:arch] ||= module_info['AdaptedArch']
+    super
+  end
+
+  def generate_payload_uuid(conf = {})
+    conf[:arch] ||= module_info['AdaptedArch']
+    conf[:platform] ||= module_info['AdaptedPlatform']
+    super
+  end
+
+  def handle_connection(conn, opts = {})
+    opts[:arch] ||= module_info['AdaptedArch']
+    super
+  end
+
+  def srvhost
+    Rex::Socket.is_ipv6?(datastore['FETCH_SRVHOST']) ? "[#{datastore['FETCH_SRVHOST']}]" : datastore['FETCH_SRVHOST']
+  end
```

Comment:
The brackets shouldn't be added here since it should always be the IP address. I would recommend the following.

1. Put this back so it's always the IP address

```suggestion
  def srvhost
    datastore['FETCH_SRVHOST']
  end
```

2. Add a new method called `#srvnetloc` (short for server network location) which is always the IP address and port. There's already a little bit of a precedence for this naming convention.
  * https://github.com/rapid7/metasploit-framework/blob/3c0222d7d0cdde4ff8aebb74672624964988e038/lib/msf/core/exploit/remote/http_server.rb#L161
  * https://github.com/rapid7/metasploit-framework/blob/3c0222d7d0cdde4ff8aebb74672624964988e038/lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/net.rb#L682

```ruby
  def srvnetloc
    netloc = srvhost
    if Rex::Socket.is_ipv6?(netloc)
      netloc = "[#{netloc}]:#{srvport}"
    else
      netloc = "#{netloc}:#{srvport}"
    end
    netloc
  end
```

3. Replacing all instances of `#{srvhost}:#{srvport}` with `#{srvnetloc}`

Reply:
Done